src/main/java/com/hivemq/persistence/local/memory/RetainedMessageMemoryLocalPersistence.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2019 dc-square GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.persistence.local.memory;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Predicate;
+import com.hivemq.annotations.ExecuteInSingleWriter;
+import com.hivemq.configuration.service.InternalConfigurations;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.annotations.Nullable;
+import com.hivemq.metrics.HiveMQMetrics;
+import com.hivemq.persistence.RetainedMessage;
+import com.hivemq.persistence.local.xodus.PublishTopicTree;
+import com.hivemq.persistence.payload.PublishPayloadPersistence;
+import com.hivemq.persistence.retained.RetainedMessageLocalPersistence;
+import com.hivemq.util.PublishUtil;
+import com.hivemq.util.ThreadPreConditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.hivemq.util.ThreadPreConditions.SINGLE_WRITER_THREAD_PREFIX;
+
+/**
+ * @author Lukas Brandl
+ */
+@Singleton
+public class RetainedMessageMemoryLocalPersistence implements RetainedMessageLocalPersistence {
+
+    private static final Logger log = LoggerFactory.getLogger(RetainedMessageMemoryLocalPersistence.class);
+
+    @VisibleForTesting
+    final @NotNull AtomicLong currentMemorySize = new AtomicLong();
+
+    @VisibleForTesting
+    @NotNull
+    final ConcurrentHashMap<Integer, PublishTopicTree> topicTrees = new ConcurrentHashMap<>();
+
+    private final @NotNull ConcurrentHashMap<Integer, Map<String, RetainedMessage>> buckets = new ConcurrentHashMap<>();
```

Comment:
This can be an array as we already build our own (fixed size) hash table.

Reply:
Generic arrays are not possible but i can use a immutable list here

Reply:
Just use a raw type for the initialization. Afterwards its type safe.
`
        //noinspection unchecked
        buckets = new ConcurrentHashMap[bucketCount];
        for (int i = 0; i < bucketCount; i++) {
            buckets[i] = new ConcurrentHashMap<>();
        }
`